src/MCPClient/lib/clientScripts/verify_checksum.py

```diff
@@ -0,0 +1,276 @@
+        "metadata/checksum.md5": "md5sum",
+        "metadata/checksum.sha1": "sha1sum",
+        "metadata/checksum.sha256": "sha256sum",
+        "metadata/checksum.sha512": "sha512sum",
+        "metadata/checksum.b2": "b2sum",
+    }
+
+    OKAY_STRING = ": OK"
+    FAIL_STRING = ": FAILED"
+    ZERO_STRING = "no properly formatted"
+    IMPROPER_STRING = "improperly formatted"
+    FAILED_OPEN = ": FAILED open or read"
+    EXIT_NON_ZERO = "returned non-zero exit status 1"
+
+    def __init__(self, path, job=print):
+        """Class constructor."""
+        try:
+            self.COMMAND = self.COMMANDS[path]
+            self.job = job
+            self.command_called = None
+        except KeyError:
+            raise NoHashCommandAvailable()
+
+    def _call(self, *args):
+        """Make the call to Python subprocess and record the command being
+        called.
+        """
+        self.command_called = (self.COMMAND,) + args
+        return self._decode(subprocess.check_output(self.command_called))
+
+    def count_and_compare_lines(self, hashfile, objectsdir):
+        """Count the number of lines in a checksum file and compare with the
+        number of objects being transferred. The requirement of hashsum as
+        this microservice job is written is that the mapping is 1:1. There
+        isn't space for an empty line at the end of the file.
+        """
+        lines = self._count_lines(hashfile)
+        objects = self._count_files(objectsdir)
+        if lines == objects:
+            return True
+        self.job.pyprint(
+            "{}: Comparison failed with {} checksum lines and {} "
+            "transfer files".format(self.get_ext(hashfile), lines, objects),
+            file=sys.stderr,
+        )
+        return False
+
+    def compare_hashes(self, hashfile, objectsdir):
+        """Compare transfer files with the checksum file provided."""
+        if not self.count_and_compare_lines(hashfile, objectsdir):
+            return 1
+        try:
+            self._call("-c", "--strict", hashfile)
+            return 0
+        except subprocess.CalledProcessError as err:
+            if self.EXIT_NON_ZERO in str(err):
+                warn = "{}: comparison exited with status: {}. Please check the formatting of the checksums or integrity of the files.".format(
+                    self.get_ext(hashfile), err.returncode
+                )
+                self.job.pyprint(warn, file=sys.stderr)
+            for line in self._decode(err.output):
+                if not line:
+                    continue
+                if line.endswith(self.OKAY_STRING):
+                    continue
+                if (
+                    line.endswith(self.FAIL_STRING)
+                    or self.ZERO_STRING in line
+                    or self.IMPROPER_STRING in line
+                ):
+                    self.job.pyprint(
+                        "{}: {}".format(self.get_ext(hashfile), line), file=sys.stderr
```

Comment:
I haven't run the script so I don't know how these conditions work together, but if the `EXIT_NON_ZERO` warning doesn't always come before a message like this it could be difficult to track the issue back to the `verify_checksum` script just from the logs. Maybe it's worth adding some more context to this message.

Reply:
I might not quite understand but we can talk more. 

I have created some [sample data](https://github.com/artefactual/archivematica-sampledata/tree/ed7fc9ae8361308f3215cbd32f970bea8dd58509/training-resources/fixityCheckShouldFail) to specifically test failures, the output on `stderr` for each of the checksums for different failures will look as follows:
```
Comparing transfer checksums with the supplied sha512 file
sha512: comparison exited with status: 1. Please check the formatting of the checksums or integrity of the files.
Comparing transfer checksums with the supplied sha1 file
sha1: Comparison failed with 2 checksum lines and 3 transfer files
Comparing transfer checksums with the supplied b2 file
b2: comparison exited with status: 1. Please check the formatting of the checksums or integrity of the files.
b2: objects/Nemastylis_geminiflora_Flower.PNG: FAILED
b2: objects/oakland03.jp2: FAILED
Comparing transfer checksums with the supplied md5 file
md5: comparison exited with status: 1. Please check the formatting of the checksums or integrity of the files.
md5: objects/Nemastylis_geminiflora_Flower.PNG: FAILED
md5: objects/oakland03.jp2: FAILED
md5: WFPC01.GIF: FAILED open or read
Comparing transfer checksums with the supplied sha256 file
sha256: Comparison failed with 1 checksum lines and 3 transfer files
```
So on the one hand I think we've covered a lot. But it would be good to understand a bit more about the context that might be possible to add?